### PR TITLE
Add redundant bl grouping

### PR DESCRIPTION
**Pull request by nkern**
for taking cross-spectra within redundant baseline groups

### CODE REVIEW

File: hera_pspec/pspecdata.py [hunk 1]
```diff
@@ -727,42 +760,78 @@ def scalar(self, stokes='I', taper='none', little_h=True, num_steps=2000):
...
+            In this case, pspec will do one of two things:
+
+                1) The data in each baseline-group are averaged together before squaring (avg_groups=True),
+                   and then crossed with the same positional baseline group between bls1 and bls2.
+
+                2) All permutations of cross-spectra between bls in a group in bls1 and bls in a group
+                   in bls2 are calculated (avg_groups=False). Note that baselines are never crossed between
+                   different baseline groups, which are defined based on their positional index.
+
+                   If exclude_auto_bls == True, bl-pairs that have a repeated baseline are eliminted.
```
Comment (philbull) [inline hera_pspec/pspecdata.py]: Typo.
Also, can this be expanded a bit to include an example, and to define what happens with groups? What happens when `avg_groups=True` and the two groups contain the same baseline?

File: hera_pspec/pspecdata.py [hunk 2]
```diff
@@ -727,42 +760,78 @@ def scalar(self, stokes='I', taper='none', little_h=True, num_steps=2000):
...
+                   If exclude_conjugated_blpairs == True, if a baseline pair exists as well as its conjugate, 
+                   the latter is eliminated. Ex: ((1, 2), (2, 3)) and ((2, 3), (1, 2))
```
Comment (philbull) [inline hera_pspec/pspecdata.py]: The example is a little unclear, as the user might not understand the tuple notation. `((1,2), (2,3))` actually means dataset1(1,2) x dataset2(2,3) doesn't it?

File: hera_pspec/pspecdata.py [hunk 3]
```diff
@@ -775,74 +844,152 @@ def pspec(self, bls, beam=None, input_data_weight='identity', norm='I',
...
 
+        exclude_conjugated_blpairs : boolean, optional
+            If bls1 and bls2 are lists of bl groups, exclude conjugated baseline-pairs.
+            Ex. If ((1, 2), (2,3)) and ((2, 3), (1,2)) exist, exclude the latter.
+
+        exclude_auto_bls : boolean, optional
+            If bls1 and bls2 are lists of bl groups, exclude bl-pairs when a bl is paired with itself.
+
+        avg_groups : boolean, optional
+            If bls1 and bls2 are a list of bl groups, average data in each group before squaring.
```
Comment (philbull) [inline hera_pspec/pspecdata.py]: "before squaring" -> "before cross-multiplying"?

File: hera_pspec/pspecdata.py [hunk 4]
```diff
@@ -775,74 +844,152 @@ def pspec(self, bls, beam=None, input_data_weight='identity', norm='I',
...
+            fed_bl_group = False
+
+        elif isinstance(bls1[0], list) and isinstance(bls1[0][0], tuple) and isinstance(bls2[0], list) \
+            and isinstance(bls2[0][0], tuple):
+            # bls1 and bls2 fed as list of bl groups
+            fed_bl_group = True
+            assert len(bls1) == len(bls2), "if fed as list of bl groups, len(bls1) must equal len(bls2)"
+
+        else:
+            raise TypeError("bls1 and bls2 must both be fed as either a list of bl tuples, or a list of bl groups")
```
Comment (philbull) [inline hera_pspec/pspecdata.py]: Does this mean that we don't support mixed lists, like `bls1 = [[tuple, tuple], tuple, tuple]`?

**Comment by nkern** [comment]
thanks @philbull I'm going to merge this as is now, and add your comments to the `UVPSpec` PR, b/c that branch already has this one merged in...

File: hera_pspec/pspecdata.py [hunk 5]
```diff
@@ -727,42 +760,78 @@ def scalar(self, stokes='I', taper='none', little_h=True, num_steps=2000):
...
 
         return scalar
 
-    def pspec(self, bls, beam=None, input_data_weight='identity', norm='I', 
-              taper='none', little_h=True, verbose=True):
+    def pspec(self, bls1, bls2, dset1_ind, dset2_ind, input_data_weight='identity', norm='I', taper='none', 
+              little_h=True, avg_groups=False, exclude_auto_bls=False, exclude_conjugated_blpairs=False,
+              verbose=True):
         """
         Estimate the delay power spectrum from the datasets contained in this
```
[the remark this answers is not in the document]
Comment (nkern) [reply hera_pspec/pspecdata.py]: good catch

File: hera_pspec/pspecdata.py [hunk 6]
```diff
@@ -130,6 +130,39 @@ def validate_datasets(self, verbose=True):
...
+    def check_key_in_dset(self, key, dset_ind):
+        """
+        Check 'key' exists in the UVData object self.dsets[dset_ind]
+
+        Parameters
+        ----------
+        key : tuple
+            if length 1: assumed to be polarization number or string
+            elif length 2: assumed to be antenna-number tuple (ant1, ant2)
+            elif length 3: assuemd ot be antenna-number-polarization tuple (ant1, ant2, pol)
```
[the remark this answers is not in the document]
Comment (nkern) [reply hera_pspec/pspecdata.py]: thanks

File: hera_pspec/pspecdata.py [hunk 7]
```diff
@@ -727,42 +760,78 @@ def scalar(self, stokes='I', taper='none', little_h=True, num_steps=2000):
...
+                                                            taper=taper, little_h=little_h, num_steps=num_steps)
+        else:
+            scalar = beam.compute_pspec_scalar(self.freqs[0], self.freqs[-1], len(self.freqs), stokes=stokes,
+                                                            taper=taper, little_h=little_h, num_steps=num_steps)
 
         return scalar
 
-    def pspec(self, bls, beam=None, input_data_weight='identity', norm='I', 
-              taper='none', little_h=True, verbose=True):
+    def pspec(self, bls1, bls2, dset1_ind, dset2_ind, input_data_weight='identity', norm='I', taper='none',
```
[the remark this answers is not in the document]
Comment (nkern) [reply hera_pspec/pspecdata.py]: yeah we can do that, make sit a little more succinct

File: hera_pspec/pspecdata.py [hunk 8]
```diff
@@ -727,42 +760,78 @@ def scalar(self, stokes='I', taper='none', little_h=True, num_steps=2000):
...
+            Alternatively, bls1 and bls2 can contain lists of baselines groups, which are themselves lists of
+            baseline tuples. A bl-group in bls1 should have the same positional index in bls2.
+            In this case, pspec will do one of two things:
+
+                1) The data in each baseline-group are averaged together before squaring (avg_groups=True),
+                   and then crossed with the same positional baseline group between bls1 and bls2.
+
+                2) All permutations of cross-spectra between bls in a group in bls1 and bls in a group
+                   in bls2 are calculated (avg_groups=False). Note that baselines are never crossed between
+                   different baseline groups, which are defined based on their positional index.
```
[the remark this answers is not in the document]
Comment (nkern) [reply hera_pspec/pspecdata.py]: sounds good.

File: hera_pspec/pspecdata.py [hunk 9]
```diff
@@ -727,42 +760,78 @@ def scalar(self, stokes='I', taper='none', little_h=True, num_steps=2000):
...
+            "right-hand" dataset. A baseline tuple is specified as an antenna pair, Ex: (ant1, ant2)
+
+            Alternatively, bls1 and bls2 can contain lists of baselines groups, which are themselves lists of
+            baseline tuples. A bl-group in bls1 should have the same positional index in bls2.
+            In this case, pspec will do one of two things:
+
+                1) The data in each baseline-group are averaged together before squaring (avg_groups=True),
+                   and then crossed with the same positional baseline group between bls1 and bls2.
+
+                2) All permutations of cross-spectra between bls in a group in bls1 and bls in a group
```
[the remark this answers is not in the document]
Comment (nkern) [reply hera_pspec/pspecdata.py]: yup, thanks

File: hera_pspec/pspecdata.py [hunk 10]
```diff
@@ -727,42 +760,78 @@ def scalar(self, stokes='I', taper='none', little_h=True, num_steps=2000):
...
         object, using the optimal quadratic estimator from arXiv:1502.06016.
 
+        In this formulation, the power spectrum is proportional to the visibility data via
+
+        P = M data_{LH} E data_{RH}
+
+        where E contains the data weighting and FT matrices, M is a normalization matrix, 
+        and the two separate datasets are denoted as "left-hand" and "right-hand".
+
+        Each power spectrum is generated by taking a bl from bls1 out of the the dset1_ind'th dataset
```
[the remark this answers is not in the document]
Comment (nkern) [reply hera_pspec/pspecdata.py]: yup

File: hera_pspec/pspecdata.py (hunk 1 above)
Comment (nkern) [reply hera_pspec/pspecdata.py]: yup! i can add an example for `exclude_auto_bls`. 

ah, average groups doesn't average between groups, it just means to average within a group.. perhaps that should be made more clear..

Comment (nkern) [reply hera_pspec/pspecdata.py]: i'm going to change the keywarg to `avg_group`, which i think is a little truer to what it's doing (even though there may be multiple groups, each of which get averaged independently)

File: hera_pspec/pspecdata.py (hunk 2 above)
Comment (nkern) [reply hera_pspec/pspecdata.py]: yeah it does. updated the doc at the top of the string to explain what a "baseline-pair" is..

File: hera_pspec/pspecdata.py (hunk 4 above)
Comment (nkern) [reply hera_pspec/pspecdata.py]: the answer is yes, `bls1` should be fed as either `[tuple, tuple, ..., tuple]` or `[[tuple, tuple, ...], [tuple, ...], ..., [tuple]]`, but what you describe above can be achieved by wrapping a lone-tuple within a list. its up to the user to know to do that (and also `redcal` will automatically do that for you anyways)

**Pull request merged**